### PR TITLE
Allow object type form filed values to be Exported into CSV 

### PR DESCRIPTION
**Pull request by patelswapnil**
This PR adds following abilities:
 - Loop through the object type form filed values to include it in CSV by transforming each key of the object into header and ordering the values under their relevant header (key)
 -  to set a function to `exportTransform` key to fields in `form` DSL to return the the object (can be embedded Object as well) which will be exported into CSV.

__Notes:__ 
  - Each key of the object will become a header (column label) in CSV and its value will appear under that header.
  - Any `key` which has `object` as a value (in case of embedded object), will not be included in headers in exported CSV. 

### Usage

  - A project that implements `linz` can provide an `exportTransform` method  for a form field to process the value and return an `object` in specific format. This object will be looped through by `linz` to convert that object into CSV format:

```
exportTransform: function (fieldValue) {
     
     // process the value, create an object as required and return it 
     let obj = {
          key1: 'value of key1',
          embeddedKey: {
                eKey1: 'value of eKey1',
                eKey2: {
                      k1: [1, 2, 3],
                      k2: 'value of k2 of eKey2'
                }
          }
     };

     return obj;

}
```

The  exported CSV will be as follow:

```
Field1, Field2, ......., key1         , eKey1         , k1     , k2
value1, value2, ......., value of key1, value of eKey1, "1,2,3", value of k2 of eKey2
```

__Note:__ Keys, `embeddedKey` and `eKey2` are not included in exported CSV.

### CODE REVIEW

**Comment by patelswapnil** [comment]
@smebberson this PR is ready for review.

File: lib/utils.js [hunk 1]
```diff
@@ -95,7 +102,11 @@ exports.json2CSV = function (obj) {
 
     });
 
-    return '"' + data.join('","') + '"\n';
+    if (labels.length) {
+        str = labels.join() + '\n';
```
Comment (smebberson) [inline lib/utils.js]: Please rewrite using es6 template strings (i.e. `${labels.join()}\n`).

File: lib/utils.js [hunk 2]
```diff
@@ -95,7 +102,11 @@ exports.json2CSV = function (obj) {
...
+    }
+
+    return str + '"' + data.join('","') + '"\n';
```
Comment (smebberson) [inline lib/utils.js]: Please rewrite using es6 template strings, I think it would be much easier to read.

File: middleware/modelExport.js [hunk 3]
```diff
@@ -70,10 +61,86 @@ var modelExportHelpers = function modelExportHelpers (req, res) {
             return (linz.utils.asBoolean(val) ? 'Yes' : 'No');
         }
 
+        if (typeof val === 'object') {
+
+            if (!fieldName) {
+                return val;
+            }
+
+            if (!Model.linz.formtools.form[fieldName].exportTransform) {
```
Comment (smebberson) [inline middleware/modelExport.js]: Can this go up in the previous `if` statement?

File: middleware/modelExport.js [hunk 4]
```diff
@@ -70,10 +61,86 @@ var modelExportHelpers = function modelExportHelpers (req, res) {
...
+
         return val;
 
     };
 
+    var getData = function (obj) {
+
+        let keyVal = {};
+
+        function recur(data) {
```
Comment (smebberson) [inline middleware/modelExport.js]: Please add a comment as to what this function actually does...

File: middleware/modelExport.js [hunk 5]
```diff
@@ -70,10 +61,86 @@ var modelExportHelpers = function modelExportHelpers (req, res) {
...
+
+            return keyVal;
+
+        }
+
+        return recur(obj);
+
+    };
+
+    var getRecordData = function getRecordData (fields, record, form, recordIndex) {
```
Comment (smebberson) [inline middleware/modelExport.js]: Please add a comment about what this function does.

File: middleware/modelExport.js [hunk 6]
```diff
@@ -70,10 +61,86 @@ var modelExportHelpers = function modelExportHelpers (req, res) {
...
+            // val is an object, iterate through object to add each key and its value (including embedded) to orderedRecord
+            let data = getData(val);
+
+            Object.keys(data).forEach(function(key) {
+                orderedRecord[key] = data[key];
+            });
+
+        });
+
+        // pass form to json2CSV method to add headers only when json2CSV is called firt time
```
Comment (smebberson) [inline middleware/modelExport.js]: Typo...

File: middleware/modelExport.js [hunk 7]
```diff
@@ -86,21 +153,14 @@ var modelExportHelpers = function modelExportHelpers (req, res) {
...
 
-                var arr = [];
-
-                // get field labels
-                fields.forEach(function (fieldName) {
-                    arr.push(form[fieldName].label);
-                });
-
-                str = arr.join() + '\n';
+                let recordData = getRecordData(fields, record, form, count);
```
Comment (smebberson) [inline middleware/modelExport.js]: Is there any requirement to create a variable? Can we just use `getRecordData` inline as it was previously?

File: lib/utils.js (hunk 1 above)
Comment (patelswapnil) [inline lib/utils.js]: @smebberson , Done.

File: lib/utils.js (hunk 2 above)
Comment (patelswapnil) [inline lib/utils.js]: done.

File: middleware/modelExport.js (hunk 3 above)
Comment (patelswapnil) [inline middleware/modelExport.js]: Done.

File: middleware/modelExport.js (hunk 4 above)
Comment (patelswapnil) [inline middleware/modelExport.js]: Done.

File: middleware/modelExport.js (hunk 5 above)
Comment (patelswapnil) [inline middleware/modelExport.js]: Done.

File: middleware/modelExport.js (hunk 6 above)
Comment (patelswapnil) [inline middleware/modelExport.js]: Fixed the typo.

File: middleware/modelExport.js (hunk 7 above)
Comment (patelswapnil) [inline middleware/modelExport.js]: Improved the code to make only one call to `getRecordData` function. Removed the `str` variable altogether as it is no longer required.

**Comment by patelswapnil** [comment]
@smebberson, this PR is ready for follow-up review.

**Comment by smebberson** [comment]
@patelswapnil, could you please alter this PR to merge into [this branch](https://github.com/linzjs/linz/tree/version-0.7.0.2), rather than master?

**Comment by allanchau** [comment]
Closing this in favour of the new PR.

**Pull request closed without merging**